src/assets/TokensController.ts

```diff
@@ -505,66 +647,120 @@ export class TokensController extends BaseController<
    * @returns The updated `allTokens` and `allIgnoredTokens` state.
    */
-  _getNewAllTokensState(newTokens: Token[], newIgnoredTokens: string[]) {
-    const { allTokens, allIgnoredTokens } = this.state;
+  _getNewAllTokensState(params: {
+    newTokens?: Token[];
+    newIgnoredTokens?: string[];
+    newDetectedTokens?: Token[];
+  }) {
+    const { newTokens, newIgnoredTokens, newDetectedTokens } = params;
+    const { allTokens, allIgnoredTokens, allDetectedTokens } = this.state;
     const { chainId, selectedAddress } = this.config;
-    const networkTokens = allTokens[chainId];
-    const networkIgnoredTokens = allIgnoredTokens[chainId];
 
-    const newNetworkTokens = {
-      ...networkTokens,
-      ...{ [selectedAddress]: newTokens },
-    };
-    const newIgnoredNetworkTokens = {
-      ...networkIgnoredTokens,
-      ...{ [selectedAddress]: newIgnoredTokens },
-    };
+    let newAllTokens = allTokens;
+    if (newTokens) {
+      const networkTokens = allTokens[chainId];
+      const newNetworkTokens = {
+        ...networkTokens,
+        ...{ [selectedAddress]: newTokens },
+      };
+      newAllTokens = {
+        ...allTokens,
+        ...{ [chainId]: newNetworkTokens },
+      };
+    }
 
-    const newAllTokens = {
-      ...allTokens,
-      ...{ [chainId]: newNetworkTokens },
-    };
+    let newAllIgnoredTokens = allIgnoredTokens;
+    if (newIgnoredTokens) {
+      const networkIgnoredTokens = allIgnoredTokens[chainId];
+      const newIgnoredNetworkTokens = {
+        ...networkIgnoredTokens,
+        ...{ [selectedAddress]: newIgnoredTokens },
+      };
+      newAllIgnoredTokens = {
+        ...allIgnoredTokens,
+        ...{ [chainId]: newIgnoredNetworkTokens },
+      };
+    }
 
-    const newAllIgnoredTokens = {
-      ...allIgnoredTokens,
-      ...{ [chainId]: newIgnoredNetworkTokens },
-    };
-    return { newAllTokens, newAllIgnoredTokens };
+    let newAllDetectedTokens = allDetectedTokens;
+    if (newDetectedTokens) {
+      const networkDetectedTokens = allDetectedTokens[chainId];
+      const newDetectedNetworkTokens = {
+        ...networkDetectedTokens,
+        ...{ [selectedAddress]: newDetectedTokens },
+      };
+      newAllDetectedTokens = {
+        ...allDetectedTokens,
+        ...{ [chainId]: newDetectedNetworkTokens },
+      };
+    }
+    return { newAllTokens, newAllIgnoredTokens, newAllDetectedTokens };
+  }
+
+  /**
+   * Fetch metadata for a token.
+   *
+   * @param tokenAddress - The address of the token.
+   * @returns The token metadata.
+   */
+  async fetchTokenMetadata(tokenAddress: string): Promise<TokenListToken> {
+    try {
+      const token = await fetchTokenMetadata<TokenListToken>(
+        isHexString(this.config.chainId)
+          ? convertHexToDecimal(this.config.chainId).toString()
+          : this.config.chainId,
+        tokenAddress,
+        this.abortController.signal,
+      );
+      return token;
+    } catch {
+      return {} as TokenListToken;
```

Comment:
This seems wrong. All of the properties of `TokenListToken` are required, but none are present here.

Reply:
Updated this method to return null within catch.